[ticket/17086] Add GitHub Codespaces support to phpBB

battye opened the pull request:
PHPBB3-17086

Checklist:

- [ ] Correct branch: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) for new features; [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html) for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: master and 3.3.x
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)
Tracker ticket (set the ticket ID to your ticket ID):

https://tracker.phpbb.com/browse/PHPBB3-17086

---

## Background

_**TLDR:** By using GitHub Codespaces, we can give all developers (that would like to use it) an identical development environment that is completely pre-configured with a LAMP stack, Xdebug, phpBB installation and VS Code -- all hosted in the cloud with no manual setup involved! The IDE and the phpBB installation are hosted entirely on GitHub's servers, and it only needs to run while it's being used. This is a massive time-saver because it can remove the need for a local development environment, particularly useful for new contributors or people who are not making big changes._

I'm a big proponent of GitHub Codespaces because I see this as the future of development and I love how they enable developers to create a fully functional environment in the cloud. I've tried Codespaces with a phpBB extension and a Symfony project and I'm blown away by how well it works.

I've added this to the phpBB project because Codespaces will allow new and existing contributors to easily jump in to the project, to make changes and tinker without being discouraged by the hassle of setting up a local environment. The cloud environment means it can be done from any computer, not just a dedicated development machine, because it's using GitHub's resources.

Most importantly, this enables Xdebug out of the box. Traditionally this has been a major hassle (think of Vagrant, Docker or WAMP) because it's a fiddly thing to get working. With Codespaces, everyone that uses it has an identical development environment so once it is set up once, it will work for everyone.

![Screenshot 2022-12-17 at 11 30 29 am](https://user-images.githubusercontent.com/2110222/208222132-d194cf35-3c74-4d21-ad7c-f7ba8ffdda6b.png)
_VS Code in debug mode in Codespaces, with the newly created phpBB board open in the second tab._

## How does it work?

Codespaces uses Docker in the backend, so between a `Dockerfile` and a `devcontainer.json` we can supply the information required for GitHub to know what sort of environment we want. In this case, a plain Ubuntu 22.04 setup is fine. Then, we install some additional things like PHP, MySQL, Apache, Xdebug and Composer before using the phpBB CLI tool to install phpBB.

By choosing to expose port 80 (and 9003 for debugging), we can access the web server running phpBB to test the code. This could be a really good way to test changes from Pull Requests in a consistent and predictable environment.

For more instructions on how to try running phpBB on Codespaces, please read my [codespaces.md](https://github.com/battye/phpbb/blob/49d39c8389c4085e649f9a6eb904cddfa692ecc6/phpBB/docs/codespaces.md) document.

## Notes

We have `.vscode/` in `.gitignore` for obvious reasons, we don't want people committing IDE configurations. However, I have included it here because I see a benefit in having Codespaces automatically pick up the path mapping for debugging. It's not strictly necessary, because people can manually add these configurations in to their Codespaces environment but it would mean debugging doesn't work out of the box. I'd be keen to hear people's thoughts about this.

Finally, Codespaces won't be suitable for everyone and it's not intended to replace a local development environment (VS Code is a fine product with many extensions but it is not built for PHP development in the same way that PhpStorm is). It is however a very good tool which can be useful to people, and I see no harm in us making it available for people to use in the phpBB repository.

iMattPro commented:
Maybe update the README too, under the contributing section, to link to the codespaces.md. Maybe update the section that talks about using Vagrant into something like a list of the now 2 options for virtualized development environments.

marc1706 commented on .devcontainer/resources/phpbb-config.yml:
```diff
@@ -0,0 +1,38 @@
+installer:
+    admin:
+        name: admin
+        password: mypassword
```
Maybe just `adminadmin` similar to the default user in unit tests might make sense here.

marc1706 commented:
Generally a very nice feature, thanks for putting in the time and preparing this. I think it's fine to create an appropriate ticket for it in the tracker even though it doesn't directly phpBB per se. If you want I can also create the ticket and adjust the commits for it.

battye commented:
Thanks @iMattPro and @marc1706, I've updated the doc and changed the default password to adminadmin.
I've also created a ticket at https://tracker.phpbb.com/browse/PHPBB3-17086 and adjusted the old commit messages.

Should I point this PR to the 3.3.x branch instead of master?

marc1706 commented:
If it works fine with 3.3.x then I'd opt for also adding it there.
